doschema/jsvalidation.py

```diff
@@ -0,0 +1,316 @@
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+import logging
+
+import jsonschema
+import six
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException, JSONSchemaEnumException, \
+    JSONSchemaRootNotObjectException, NotSupportedException
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False,
+                 allow_none_type=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver_factory: Resolver used to retrieve referenced schemas.
+                                 If not provided it will use default.
+        :param allow_none_type:
+        """
+        self.ignore_index = ignore_index
+        self.allow_none_type = allow_none_type
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        if (schema.get('type') != 'object' and 'properties' not in schema and
+                'dependencies' not in schema):
+            err_msg = "Root field / type is not 'object' in {0}"
+            raise JSONSchemaCompatibilityException(
+                '/',
+                self.uri,
+                err_msg.format(self.uri)
+            )
+
+        self.validate_schema(schema, ())
+
+        for path, field in six.iteritems(self.fields_types_dict):
+            if field.field_type == "null":
+                logging.warning(
+                    'Null type in field ' + self.make_json_pointer(
+                            field.field_tuple
+                        )
+                )
+
+        self.uri = None
+        self.resolver = None
+
+    def validate_schema(self, curr_schema, curr_field):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        self.validate_collection(curr_schema, curr_field)
+
+        field = self.fields_types_dict.get(curr_field)
+
+        if "type" in curr_schema:
+            if field:
+                if field.field_type == "null":
```

Comment:
As said IRL. We won't handle "null" types for now as they are not used by known invenio-based services. We can raise `NotImplementedError` when the `type` is set to `"null"`